pennylane/decomposition/symbolic_decomposition.py

```diff
@@ -0,0 +1,202 @@
+    return {resource_rep(base_class, **base_params): 1}
+
+
+@register_resources(_adjoint_adjoint_resource)
+def adjoint_adjoint_decomp(*_, wires, base):  # pylint: disable=unused-argument
+    """Decompose the adjoint of the adjoint of a gate."""
+    base.base._unflatten(*base.base._flatten())  # pylint: disable=protected-access
+
+
+def _adjoint_controlled_resource(base_class, base_params):
+    """Resources of the adjoint of a controlled gate whose base has adjoint."""
+
+    num_control_wires = base_params["num_control_wires"]
+    controlled_base_class = base_params["base_class"]
+
+    # Handle controlled-X gates, the adjoint is just themselves
+    if controlled_base_class is qml.X:
+        if num_control_wires == 1:
+            return {resource_rep(qml.CNOT): 1}
+        if num_control_wires == 2:
+            return {resource_rep(qml.Toffoli): 1}
+        return {
+            resource_rep(
+                qml.MultiControlledX,
+                num_control_wires=num_control_wires,
+                num_zero_control_values=base_params["num_zero_control_values"],
+                num_work_wires=base_params["num_work_wires"],
+            ): 1
+        }
+
+    # Handle custom controlled gates. The adjoint of a general controlled operator that
+    # is equivalent to a custom controlled operator should just be the custom controlled
+    # operator given that its base has_adjoint.
+    if (controlled_base_class, num_control_wires) in base_to_custom_ctrl_op():
+        custom_op_type = base_to_custom_ctrl_op()[(controlled_base_class, num_control_wires)]
+        # All gates in base_to_custom_ctrl_op do not have resource params.
+        return {resource_rep(custom_op_type): 1}
+
+    # Handle the general case, here we assume that the adjoint of a controlled gate
+    # whose base has an adjoint that is of its own type, should have the same resource
+    # rep as the controlled gate itself. For example, Adjoint(Controlled(O)) should
+    # have the same resources as Controlled(O) if the adjoint of O is another O.
+    return {resource_rep(base_class, **base_params): 1}
+
+
+@register_resources(_adjoint_controlled_resource)
+def adjoint_controlled_decomp(*_, base, **__):
+    """Decompose the adjoint of a controlled gate whose base has adjoint.
+
+    Precondition:
+    - isinstance(base, qml.ops.Controlled) and base.base.has_adjoint
+
+    """
+    qml.ctrl(
+        base.base.adjoint(),
+        control=base.control_wires,
+        control_values=base.control_values,
+        work_wires=base.work_wires,
+    )
+
+
+@functools.lru_cache()
+def has_adjoint_ops():
+    """A set of operators whose adjoint is an instance of its own type."""
+    return {
+        # identity
+        qml.Identity,
+        qml.GlobalPhase,
+        # non-parametric gates
+        qml.H,
+        qml.X,
+        qml.Y,
+        qml.Z,
+        qml.SWAP,
+        qml.ECR,
+        # single-qubit parametric gates
+        qml.Rot,
+        qml.U1,
+        qml.U2,
```

Comment:
Is the adjoint of a U2 gate again of the U2 form? :)

Reply:
I think so: https://github.com/PennyLaneAI/pennylane/blob/711375b7e526cf976e0de8addde385bbfcfca21d/pennylane/ops/qubit/parametric_ops_single_qubit.py#L994

Reply:
Ah, sorry, should've checked there myself :facepalm: Thanks!